### PR TITLE
reference: add TiDB Binlog FAQ

### PR DESCRIPTION
**Pull request by juliezhang1112**
### What is changed, added or deleted? 
- Add TiDB Binlog FAQ

### What is the related PR or file link(s)? 

- [x] Reference link(s): https://github.com/pingcap/docs-cn/pull/2010 (move the binlog folder)
- [x] This PR is to align with: https://github.com/pingcap/docs-cn/pull/1841 
- [ ] N/A (not applicable)

### Which TiDB version(s) does your changes apply to? 

- [x] All active versions: dev, v3.0, v2.1, v3.1
- [ ] dev (the latest development version)
- [ ] v3.0 (TiDB 3.0 versions)
- [ ] v2.1 (TiDB 2.1 versions)
- [ ] v3.1 (TiDB 3.1 versions)
- [ ] v4.0 (TiDB 4.0 versions)
- [ ] All active and inactive versions
- [ ] N/A (not applicable)

**Note:** If your changes apply to multiple TiDB versions, make sure you update the documents in the **corresponding** version folders such as "dev", "v3.0", "v2.1" and "v3.1" in this PR.

- [x] Updated one version first. Will update other versions after I get two LGTMs.

### CODE REVIEW

File: dev/reference/tidb-binlog/faq.md [hunk 1]
```diff
@@ -1,13 +1,139 @@
...
+* Create
+* Drop
+* Alter
+* Execute
+* Index
+* Select
+
+## What can I do if the Pump disk is almost full?
+
+If TiDB GC is normal and the **gc_tso** time in the Pump monitoring panel is identical with the configuration file, you can perform the following steps to reduce the amount of space required for a single Pump:
```
Comment (july2993) [inline dev/reference/tidb-binlog/faq.md]: s/TiDB GC/pump gc

it's pump's GC, no anything about TiDB GC

File: dev/reference/tidb-binlog/faq.md [hunk 2]
```diff
@@ -1,13 +1,139 @@
...
+
+- Modify the **GC** parameter of Pump to reduce the number of days to retain data.
+
+- Add pump pods.
```
Comment (july2993) [inline dev/reference/tidb-binlog/faq.md]: s/pods/instances

File: dev/reference/tidb-binlog/faq.md [hunk 3]
```diff
@@ -1,13 +1,139 @@
...
+
+## What can I do if Drainer synchronization is interrupted?
+
+Execute the following command to check whether the status of Pump is normal and whether all the Pump pods that are not in the `offline` state are running.
+
+{{< copyable "shell-regular" >}}
+
+``` bash
+Binlogctl -cmd pumps
```
Comment (july2993) [inline dev/reference/tidb-binlog/faq.md]: s/Binlogctl/binlogctl

File: dev/reference/tidb-binlog/faq.md [hunk 4]
```diff
@@ -1,13 +1,139 @@
...
+
+If TiDB GC is normal and the **gc_tso** time in the Pump monitoring panel is identical with the configuration file, you can perform the following steps to reduce the amount of space required for a single Pump:
+
+- Modify the **GC** parameter of Pump to reduce the number of days to retain data.
+
+- Add pump pods.
+
+## What can I do if Drainer synchronization is interrupted?
+
+Execute the following command to check whether the status of Pump is normal and whether all the Pump pods that are not in the `offline` state are running.
```
Comment (july2993) [inline dev/reference/tidb-binlog/faq.md]: s/pods/instances

File: dev/reference/tidb-binlog/faq.md [hunk 5]
```diff
@@ -1,13 +1,139 @@
...
+
+Then, check whether the Drainer monitor or log outputs corresponding errors. If so, resolve them accordingly.
+
+## What can I do if Drainer is slow to synchronizes the downstream TiDB/MySQL?
+
+Check the following monitoring items:
+
+- For the **Drainer Event** monitoring metric, check the speed of Drainer synchronizing `INSERT`, `UPDATE` and `DELETE` transactions to the downstream per second.
+
+- For the **SQL Query Time** monitoring metric, check the time Drainer takes to execute SQL statements in the downstream .
```
Comment (july2993) [inline dev/reference/tidb-binlog/faq.md]:
```suggestion
- For the **SQL Query Time** monitoring metric, check the time Drainer takes to execute SQL statements in the downstream.
```

File: dev/reference/tidb-binlog/faq.md [hunk 6]
```diff
@@ -1,13 +1,139 @@
...
+
+Possible causes and solutions for slow synchronization:
+
+- If the synchronized database contains a table without a primary key or unique index, add a primary key to the table.
+
+- If the latency between Drainer and the downstream is high, increase the value of the `worker-count` parameter of Drainer. For cross-machine synchronization, it is recommended to deploy Drainer in the downstream.
```
Comment (july2993) [inline dev/reference/tidb-binlog/faq.md]: s/cross-machine/cross-datacentor

not cross machine

File: dev/reference/tidb-binlog/faq.md [hunk 7]
```diff
@@ -1,13 +1,139 @@
...
+
+- If the load in the downstream is not high, increase the value of the `worker-count` parameter of Drainer.
+
+## What can I do if a Pump pod crashes?
```
Comment (july2993) [inline dev/reference/tidb-binlog/faq.md]: s/pod/instance

File: dev/reference/tidb-binlog/faq.md [hunk 8]
```diff
@@ -1,13 +1,139 @@
...
+
+2. To restore the latest data of the backup file, use Reparo to set `start-tso` = {snapshot timestamp of the full backup + 1} and `end-ts` = 0 (or you can specify a point in time).
+
+## How to redeploy Drainer when Master-Slave synchronization enables `ignore-error` but causes a critical error?
+
+If TiDB fails to write binlog after enabling `ignore-error`, it causes a critical error and TiDB stops writing binlog. In this situation, there is data loss of binlog. To resume synchronization, perform the following steps:
+
+1. Stop the Drainer.
+
+2. Restart the TiDB server instance that causes critical error and resume writing binlog (TiDB does not write binlog to Pump after critical error is returned).
```
Comment (july2993) [inline dev/reference/tidb-binlog/faq.md]: s/returned/triggered ?

**Comment by juliezhang1112** [comment]
Updated. PTAL again @july2993 Thanks~

File: dev/reference/tidb-binlog/faq.md [hunk 9]
```diff
@@ -1,13 +1,139 @@
...
 
 # TiDB Binlog FAQ
 
 This document collects the frequently asked questions (FAQs) about TiDB Binlog.
 
+## What is the impact of enabling TiDB Binlog on the performance of TiDB?
+
+- There is no impact on the query.
+
+- There is a slight performance impact on `INSERT`, `DELETE` and `UPDATE` transactions. In terms of latency, a p-binlog is written concurrently in the prewrite stage before committing the transactions. Generally, writing binlog is faster than the prewrite, so it does not increase latency. You can check the response time of writing binlog in Pump's monitoring panel.
```
Comment (july2993) [inline dev/reference/tidb-binlog/faq.md]: need to  emphasize faster compare with prewrite on tikv

File: dev/reference/tidb-binlog/faq.md [hunk 10]
```diff
@@ -1,13 +1,139 @@
...
+
+## What permissions does Drainer's account need to synchronize the downstream TiDB/MySQL?
+
+To synchronize the downstream TiDB/MySQL, Drainer's account requires the following permissions:
```
Comment (july2993) [inline dev/reference/tidb-binlog/faq.md]: s/permissions/privileges   ?

https://pingcap.com/docs/v2.1-legacy/sql/privilege/

File: dev/reference/tidb-binlog/faq.md [hunk 11]
```diff
@@ -1,13 +1,139 @@
...
+
+## What can I do if Drainer is slow to synchronizes the downstream TiDB/MySQL?
+
+Check the following monitoring items:
+
+- For the **Drainer Event** monitoring metric, check the speed of Drainer synchronizing `INSERT`, `UPDATE` and `DELETE` transactions to the downstream per second.
+
+- For the **SQL Query Time** monitoring metric, check the time Drainer takes to execute SQL statements in the downstream.
+
+Possible causes and solutions for slow synchronization:
```
Comment (july2993) [inline dev/reference/tidb-binlog/faq.md]: s/synchronization/replication/  ?

File: dev/reference/tidb-binlog/faq.md [hunk 12]
```diff
@@ -1,13 +1,139 @@
...
+
+- If the synchronized database contains a table without a primary key or unique index, add a primary key to the table.
+
+- If the latency between Drainer and the downstream is high, increase the value of the `worker-count` parameter of Drainer. For cross-datacenter synchronization, it is recommended to deploy Drainer in the downstream.
+
+- If the load in the downstream is not high, increase the value of the `worker-count` parameter of Drainer.
+
+## What can I do if a Pump instance crashes?
+
+If a Pump instance crashes, Drainer cannot synchronize data to the downstream because it cannot obtain the data of this instance. If this Pump instance can return to normal use, Drainer resumes synchronization; if not, perform the following steps:
```
Comment (july2993) [inline dev/reference/tidb-binlog/faq.md]: s/synchronize/replicate
s/synchronization/replication/

File: dev/reference/tidb-binlog/faq.md [hunk 13]
```diff
@@ -1,13 +1,139 @@
...
+
+1. Use [binlogctl to change the state of this Pump instance to `offline`](/dev/reference/tidb-binlog/maintain.md) to discard the data of this Pump instance.
```
Comment (july2993) [inline dev/reference/tidb-binlog/faq.md]: s/to/will ?

drainer will ignore this pump instance, the operation will not delete any data.

File: dev/reference/tidb-binlog/faq.md [hunk 14]
```diff
@@ -1,13 +1,139 @@
...
+
+    iv. Synchronize the full backup data to the downstream.
+
+    v. Deploy Drainer and use `initialCommitTs` (set `initialCommitTs` as the snapshot timestamp of the full backup) as the start point of initial synchronization.
+
+## What is checkpoint?
+
+Checkpoint records the `commit-ts` that Drainer synchronizes to the downstream. When Drainer restarts, it reads the checkpoint and then synchronizes data to the downstream starting from the corresponding `commit-ts`. The `["write save point"] [ts=411222863322546177]` Drainer log means saving the checkpoint with the corresponding timestamp.
+
+Checkpoint is saved in different files for different types of downstream:
```
Comment (july2993) [inline dev/reference/tidb-binlog/faq.md]: s/different files/different way  ?

may save in file or MySQL/TiDB  not different files ?

File: dev/reference/tidb-binlog/faq.md [hunk 15]
```diff
@@ -1,13 +1,139 @@
...
+
+- For Kafka/file, it is saved in the file of the corresponding configuration directory.
+
+The data of kafka/file contains `commit-ts`, so if the checkpoint is lost, you can check the latest `commit-ts` of the downstream data by consuming the latest data in the downstream .
+
+Drainer reads the checkpoint when it starts. If it cannot read the checkpoint, Drainer uses the configured `initialCommitTs` as the start point of the initial synchronization.
+
+## How to redeploy Drainer on the new machine when Drainer fails without affecting the data in the downstream?
+
+If the data in the downstream is not affected, you can redeploy Drainer on the new machine as long as the data can be synchronized from the corresponding checkpoint.
```
Comment (july2993) [inline dev/reference/tidb-binlog/faq.md]: s/synchronized/replicated

File: dev/reference/tidb-binlog/faq.md [hunk 16]
```diff
@@ -1,13 +1,139 @@
...
+
+1. Stop the Drainer.
+
+2. Restart the TiDB server instance that causes critical error and resume writing binlog (TiDB does not write binlog to Pump after critical error is triggered).
+
+3. Perform a full backup in the upstream.
+
+4. Clear the data in the downstream including the `tidb_binlog.checkpoint` table.
+
+5. Synchronize the full backup data to the downstream.
```
Comment (july2993) [inline dev/reference/tidb-binlog/faq.md]: s/Synchronize/Load  or Restore

File: dev/reference/tidb-binlog/faq.md (hunk 13 above)
Comment (juliezhang1112) [reply dev/reference/tidb-binlog/faq.md]: Do you suggest replacing "Use binlogctl to change the state of this Pump instance to `offline`" with "Use binlogctl will change the state of this Pump instance to `offline`"? Please make it clear. Thanks~

Comment (july2993) [reply dev/reference/tidb-binlog/faq.md]: sorry,  I mean s/to discard/will discard/,   but I think keeping `to discard` is ok now...

Comment (juliezhang1112) [reply dev/reference/tidb-binlog/faq.md]: > sorry, I mean s/to discard/will discard/, but I think keeping `to discard` is ok now...

Got it~

**Comment by juliezhang1112** [comment]
PTAL @lilin90 Thanks~

File: dev/reference/tidb-binlog/faq.md [hunk 17]
```diff
@@ -1,13 +1,139 @@
...
+
+2. Use [binlogctl to change the state of the old Drainer to `offline`](/dev/reference/tidb-binlog/maintain.md).
+
+## How to restore the data of a cluster using a full backup and a binlog backup file?
+
+1. Clean up the cluster and restore a full backup.
+
+2. To restore the latest data of the backup file, use Reparo to set `start-tso` = {snapshot timestamp of the full backup + 1} and `end-ts` = 0 (or you can specify a point in time).
+
+## How to redeploy Drainer when Master-Slave replication enables `ignore-error` but triggers a critical error?
```
Comment (lilin90) [inline dev/reference/tidb-binlog/faq.md]:
```suggestion
## How to redeploy Drainer when enabling `ignore-error` in Master-Slave replication triggers a critical error?
```

File: dev/reference/tidb-binlog/faq.md [hunk 18]
```diff
@@ -1,13 +1,139 @@
...
+
+If TiDB fails to write binlog after enabling `ignore-error` and triggers a critical error, TiDB stops writing binlog and binlog data loss occurs. To resume replication, perform the following steps:
+
+1. Stop the Drainer.
```
Comment (lilin90) [inline dev/reference/tidb-binlog/faq.md]: instance?
```suggestion
1. Stop the Drainer instance.
```

**Comment by juliezhang1112** [comment]
Updated. PTAL again @lilin90 Thanks~

File: dev/reference/tidb-binlog/faq.md [hunk 19]
```diff
@@ -1,13 +1,139 @@
...
 
 # TiDB Binlog FAQ
 
 This document collects the frequently asked questions (FAQs) about TiDB Binlog.
 
+## What is the impact of enabling TiDB Binlog on the performance of TiDB?
+
+- There is no impact on the query.
+
+- There is a slight performance impact on `INSERT`, `DELETE` and `UPDATE` transactions. In terms of latency, a p-binlog is written concurrently in the TiKV prewrite stage before committing the transactions. Generally, writing binlog is faster than TiKV prewrite, so it does not increase latency. You can check the response time of writing binlog in Pump's monitoring panel.
```
[the remark this answers is not in the document]
Comment (lilin90) [reply dev/reference/tidb-binlog/faq.md]: 👌

File: v2.1/reference/tidb-binlog/faq.md [hunk 20]
```diff
@@ -1,13 +1,143 @@
...
+* Alter
+* Execute
+* Index
+* Select
+
+## What can I do if the Pump disk is almost full?
+
+1. Check whether Pump's GC works well:
+
+    - Check whether the **gc_tso** time in Pump’s monitoring panel is identical with that of the configuration file. For Pump of version <= v2.1.14, GC works after the non-offline drainer consumes data. If there are drainer instances that are no longer in use, use binlogctl to make them offline.
```
Comment (juliezhang1112) [inline v2.1/reference/tidb-binlog/faq.md]: This line is different for different versions. PTAL @july2993 @lilin90 Thanks~

**Comment by juliezhang1112** [comment]
> LGTM. @juliezhang1112 You can update other versions now. Thanks!

Updated. PTAL, thanks~

File: v2.1/reference/tidb-binlog/faq.md (hunk 20 above)
Comment (july2993) [reply v2.1/reference/tidb-binlog/faq.md]: LGTM

**Pull request merged**